pkg/services/kubeconfig_loader.go

```diff
@@ -0,0 +1,239 @@
+		delete(k.remoteKubeClients, client)
+		k.logger.Debugf("Removed remote client for cluster: %s", client)
+	}
+
+	// Update stored hash
+	k.lastConfigHash = currentHash
+
+	kubeConfig, err := clientcmd.LoadFromFile(k.path)
+	if err != nil { // kubeConfig is nil only if the file is empty or invalid
+		k.logger.Errorf("Error loading kube config from path %s: %s", k.path, err)
+		return
+	}
+	if len(kubeConfig.Contexts) == 0 {
+		k.logger.Warnf("No contexts found in kubeconfig file: %s", k.path)
+		return
+	}
+
+	var remoteClustersName []string
+	for contextName := range kubeConfig.Contexts {
+		remoteClustersName = append(remoteClustersName, contextName)
+	}
+	k.logger.Debugf("Remote clusters found in kubeconfig: %v", remoteClustersName)
+
+	if len(remoteClustersName) > 0 {
+		var initializedClustersName []string
+		for _, cluster := range remoteClustersName {
+			remoteClusterConfig, err := BuildConfigWithContextFromFlags(cluster, k.path)
+			if err != nil {
+				k.logger.Errorf("Error building kubeconfig for cluster '%s': %s", cluster, err)
+				continue
+			}
+
+			remoteClusterConfig.Timeout = 1 * time.Second // Set a timeout for the HTTP client, maybe this should be configurable
+			var remoteClusterClient IKubeClient
+			if k.clientFactory != nil {
+				remoteClusterClient, err = k.clientFactory(remoteClusterConfig)
+			} else {
+				remoteClusterClient, err = kubernetes.NewForConfig(remoteClusterConfig)
+				remoteClusterClient = newKubeClientAdapter(remoteClusterClient.(*kubernetes.Clientset))
+			}
+			if err != nil {
+				k.logger.Errorf("Error creating remote clientset for cluster '%s': %s", cluster, err)
+				continue
+			}
+			output, err := remoteClusterClient.ServerVersion()
+			// This is a simple way to check if the server is reachable and the config is valid
+			if err != nil {
+				k.logger.Errorf("Error getting server version for cluster '%s': %s", cluster, err)
+				continue
+			}
+			k.logger.Infof("Connected to remote cluster '%s' with server version: %s", cluster, output.String())
+
+			k.remoteKubeClients[cluster] = remoteClusterClient
+			initializedClustersName = append(initializedClustersName, cluster)
+			k.logger.Debugf("Initialized remote cluster client for '%s'", cluster)
+		}
+
+		if len(initializedClustersName) > 0 {
+			k.logger.Infof("Remote clusters successfully initialized: %v", initializedClustersName)
+		}
+	}
+}
+
+// StartWatching starts watching the kubeconfig file for changes, including deletions and recreations.
+func (k *KubeconfigLoader) StartWatching(ctx context.Context) {
+	var debounceTimer *time.Timer // timer to debounce events and avoid multiple reloads
+	debounceDuration := time.Duration(500) * time.Millisecond
+	go func() {
+		_, err := os.Stat(k.path)
+		for err != nil && os.IsNotExist(err) {
+			k.logger.Warnf("Kubeconfig file does not exist yet: %s", k.path)
+			time.Sleep(2 * time.Second) // Polling interval for file existence
+			_, err = os.Stat(k.path)
```

Comment:
`err` here is not used. What would happen if the file doesn't exist after 2 seconds? Maybe a short loop with a exit? Shouldn't we break the execution in case the file ultimately doesn't exists (maybe mount volume permissions errors and something like that)?

Reply:
Until the file is created, it waits forever. I'll add some logic to not wait forever.

Reply:
Fixed! Added logic to at most wait for the maxRetries flag.